lib/Expr/ArrayExprVisitor.cpp

```diff
@@ -0,0 +1,259 @@
+    if (re->updates.root->hash() == array->hash() && width < ce.getWidth()) {
+      if (width == Expr::InvalidWidth) width = ce.getWidth();
+    }
+  }
+  return Action::doChildren();
+}
+ExprVisitor::Action IndexTransformationExprVisitor::visitMul(const MulExpr &e) {
+  if (isa<ConstantExpr>(e.getKid(0)))
+    mul = e.getKid(0);
+  else if (isa<ConstantExpr>(e.getKid(0)))
+    mul = e.getKid(1);
+  return Action::doChildren();
+}
+
+//------------------------------ PURDUE OPTIMIZATION-------------------------//
+ExprVisitor::Action ArrayReadExprVisitor::visitConcat(const ConcatExpr &ce) {
+  ReadExpr *base = ArrayExprHelper::hasOrderedReads(ce);
+  if (base) {
+    return inspectRead(ce.hash(), ce.getWidth(), *base);
+  }
+  return Action::doChildren();
+}
+ExprVisitor::Action ArrayReadExprVisitor::visitRead(const ReadExpr &re) {
+  return inspectRead(re.hash(), re.getWidth(), re);
+}
+// This method is a mess because I want to avoid looping over the UpdateList
+// values twice
+ExprVisitor::Action ArrayReadExprVisitor::inspectRead(unsigned hash,
+                                                      Expr::Width width,
+                                                      const ReadExpr &re) {
+	// pre(*): index is symbolic
+	if (!isa<ConstantExpr>(re.index)) {
+	  if (readInfo.find(&re) == readInfo.end()) {
+		if (re.updates.root->isSymbolicArray() && !re.updates.head) {
+		  return Action::doChildren();
+		}
+		if (re.updates.head) {
+		  // Check preconditions on UpdateList nodes
+		  bool hasConcreteValues = false;
+		  for (const UpdateNode *un = re.updates.head; un; un = un->next) {
+			// Symbolic case - \inv(update): index is concrete
+			  if (!isa<ConstantExpr>(un->index)) {
+				incompatible = true;
+				break;
+			  } else if (!isa<ConstantExpr>(un->value)) {
+				// We tell the optimization that there is a symbolic value,
+				// otherwise we rely on the concrete optimization procedure
+				symbolic = true;
+			  } else if (re.updates.root->isSymbolicArray() &&
+						 isa<ConstantExpr>(un->value)) {
+				// We can optimize symbolic array, but only if they have
+				// at least one concrete value
+				hasConcreteValues = true;
+			  }
+		  }
+		  // Symbolic case - if array is symbolic, then we need at least one
+		  // concrete value
+		  if (re.updates.root->isSymbolicArray()) {
+			if (hasConcreteValues)
+			  symbolic = true;
+			else
+			  incompatible = true;
+		  }
+
+		  if (incompatible)
+			return Action::skipChildren();
+		}
+		symbolic |= re.updates.root->isSymbolicArray();
+		reads.push_back(&re);
+		readInfo.insert(
+			std::pair<const ReadExpr *, std::pair<unsigned, Expr::Width> >(
+				&re, std::pair<unsigned, Expr::Width>(hash, width)));
+	  }
+	  return Action::skipChildren();
+	}
+  return Action::doChildren();
+}
+
+ExprVisitor::Action ArrayValueOptReplaceVisitor::visitConcat(const ConcatExpr &ce) {
+  std::map<unsigned, ref<Expr> >::const_iterator found = optimized.find(ce.hash());
```

Comment:
I think we don't need the ugly "> > >" syntax anymore. Also just use "auto" for boring types?